### PR TITLE
Use native features in `lib/url`

### PR DESCRIPTION
**Pull request by sgomes**
Now that Calypso has polyfilled support for the native `URL` and `URLSearchParams`, we should try to remove usage of the browserified version of node's `url`, and the `qs` library, with the ultimate goal of removing them from (client-side) Calypso.

This PR does that for `lib/url`.

#### Changes proposed in this Pull Request

* Add tests for `decodeURIIfValid` and `decodeURIComponentIfValid`
* Rewrite `lib/url` to use native `URL` and `URLSearchParams` support instead of the browserified version of node's `url`. Also replace `lodash` usage with plain JS.
* Split `lib/url` across multiple files.

#### Testing instructions

There's a lot of code throughout Calypso that depends on `lib/url`, so it's hard to offer a thorough testing strategy.

I requested the full E2E tests on this PR to provide some measure of assurance.

### CODE REVIEW

**Comment by sgomes** [comment]
Splitting up the library appears to save some bytes in most routes, notably domains: 
http://iscalypsofastyet.com/branch?branch=update/use-native-features-in-lib-url

File: client/lib/url/decode-utils.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+	toString: () => string;
+}
+
+/**
+ * Wrap decodeURI in a try / catch block to prevent `URIError` on invalid input
+ * Passing a non-string value will return an empty string.
+ * @param  encodedURI URI to attempt to decode
+ * @return            Decoded URI (or passed in value on error)
+ */
+export function decodeURIIfValid( encodedURI: string | Stringable ): TypedURL {
```
Comment (sirreal) [inline client/lib/url/decode-utils.ts]: It's really hard to make sense of this function in a typed world 🙃 

It seems to be a helper so `decodeURI` is safe and won't crash the app. In that sense, it should probably accept and argument of type `Stringable | Falsy`.

```suggestion
export function decodeURIIfValid( encodedURI: Stringable | Falsy ): TypedURL {
```

This type of refactor would be much simpler if we had type safety, but that's a long ways away. At the moment, we'll need to open up the argument type and be more careful in the body. Callers are not updated, so we can continue to expect "bad" inputs.

Opening up the argument type requires a few other changes…

File: client/lib/url/decode-utils.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	let encodedURIString: string;
+
+	if ( encodedURI as Stringable ) {
```
Comment (sirreal) [inline client/lib/url/decode-utils.ts]: This is a type assertion and won't result in any runtime type checks, this is just a falsy check `if ( encodedURI ) …`.

File: client/lib/url/decode-utils.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+		encodedURIString = encodedURI.toString();
+	} else {
+		encodedURIString = encodedURI as string;
```
Comment (sirreal) [inline client/lib/url/decode-utils.ts]: With an opThis doesn't look right, `string` will be handled in the other branch. We'll be hitting the Falsy cases here: `null | undefined | false | 0 | ''`.

File: client/lib/url/decode-utils.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+/**
+ * Wrap decodeURI in a try / catch block to prevent `URIError` on invalid input
+ * Passing a non-string value will return an empty string.
+ * @param  encodedURI URI to attempt to decode
+ * @return            Decoded URI (or passed in value on error)
+ */
+export function decodeURIIfValid( encodedURI: string | Stringable ): TypedURL {
+	let encodedURIString: string;
```
Comment (sirreal) [inline client/lib/url/decode-utils.ts]: If I'm not mistaken, all the conditional blocks preceeding the try/catch can be handled right here without type assertions.

```suggestion
	const encodedURIString: string = encodedURI ? encodedURI.toString() : '';
```

File: client/lib/url/decode-utils.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
+/**
```
Comment (sirreal) [inline client/lib/url/decode-utils.ts]:
```suggestion
/**
 * External dependencies
 */
import { Falsy } from 'utility-types'

/**
```

File: client/lib/url/decode-utils.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+/**
+ * Wrap decodeURIComponent in a try / catch block to prevent `URIError` on invalid input
+ * Passing a non-string value will return an empty string.
+ * @param  encodedURIComponent URI component to attempt to decode
+ * @return                     Decoded URI component (or passed in value on error)
+ */
+export function decodeURIComponentIfValid( encodedURIComponent: string | Stringable ): string {
+	let encodedURIComponentString: string;
```
Comment (sirreal) [inline client/lib/url/decode-utils.ts]:
```suggestion
	const encodedURIComponentString: string = encodedURIComponent ? encodedURIComponent.toString() : '';
```

File: client/lib/url/decode-utils.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+	}
+}
+
+/**
+ * Wrap decodeURIComponent in a try / catch block to prevent `URIError` on invalid input
+ * Passing a non-string value will return an empty string.
+ * @param  encodedURIComponent URI component to attempt to decode
+ * @return                     Decoded URI component (or passed in value on error)
+ */
+export function decodeURIComponentIfValid( encodedURIComponent: string | Stringable ): string {
```
Comment (sirreal) [inline client/lib/url/decode-utils.ts]:
```suggestion
export function decodeURIComponentIfValid( encodedURIComponent: Stringable | Falsy ): string {
```

File: client/lib/url/decode-utils.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
+/**
+ * Internal dependencies
+ */
+import { URL as TypedURL } from 'types';
+
+interface Stringable {
+	toString: () => string;
+}
+
```
Comment (sirreal) [inline client/lib/url/decode-utils.ts]: It's a bit tempting to make a wrapper for both of these functions given their bodies are duplicated except for the function the call on the treated input 🙂

**Review by sirreal — COMMENTED** [review]
I left notes on `decode-utils`. How do you feel about moving the functionality function by function in independent PRs? That would make review easier.

**Comment by sgomes** [comment]
@sirreal I'll close this PR and split functions out one by one, taking into account your comments. Thanks!

**Pull request closed without merging**